website_cms/models/cms_page.py

```diff
@@ -0,0 +1,660 @@
+        page_type = self.env.ref('website_cms.default_page_type')
+        return [('type_id', '=', page_type.id)]
+
+    @api.multi
+    @api.constrains('parent_id')
+    def _check_parent_id(self):
+        """Make sure we cannot have parent = self."""
+        self.ensure_one()
+        if self.parent_id and self.parent_id.id == self.id:
+            raise exceptions.ValidationError(
+                _(u'You cannot set the parent of a page '
+                  u'equal to the page itself. Page: "%s"') % self.name
+            )
+
+    @api.model
+    def _default_type_id(self):
+        page_type = self.env.ref('website_cms.default_page_type')
+        return page_type and page_type.id or False
+
+    @api.model
+    def _default_view_id(self):
+        page_view = self.env.ref('website_cms.page_default')
+        return page_view and page_view.id or False
+
+    @api.model
+    def build_public_url(self, item):
+        """Walk trough page path to build its public URL."""
+        # XXX: this could be expensive... think about it!
+        # We could store it but then we need to update
+        # all the pages below a certain parent
+        # if the parent name changes or if a parent is moved/changed.
+        # We have the same problem with path: check for comment there.
+        # In the end the url will work nonetheless because
+        # the slug contains the object id
+        # but the path in the url will be wrong.
+        # Also, not using parents name to build the path
+        # is bad because we miss the categorization
+        # provided by each parents, that's good for
+        # good URLs, and you allow ppl to not
+        # put in each page name/title a whole keyworded
+        # description of the content itself.
+        current = item
+        parts = [to_slug(current), ]
+        while current.parent_id:
+            parts.insert(0, to_slug(current.parent_id))
+            current = current.parent_id
+        public_url = '/cms/' + '/'.join(parts)
+        return public_url
+
+    @api.multi
+    def _website_url(self, field_name, arg):
+        """Override method defined by `website.published.mixin`."""
+        res = {}
+        for item in self:
+            res[item.id] = self.build_public_url(item)
+        return res
+
+    # XXX: how to update path for the whole hierarchy
+    # whenever and ancestor parent/name
+    # - no matter the level -  is updated?
+    # Right now we are supporting explicitely 3 levels,
+    # but is not nice and it limits your hierarchy
+    # We could override the write
+    # and trigger updating of path for all the items
+    # in the same path: but how to get this???
+    @api.multi
+    @api.depends('parent_id',
+                 'parent_id.parent_id',
+                 'parent_id.parent_id.parent_id',
+                 'parent_id.parent_id.parent_id.parent_id')
+    def _compute_path(self):
+        for item in self:
+            item.path = self.build_path(item)
```

Comment:
@simahawk As we talked before, it will be better to just drop all of path computing.

Simply the controller URL can be `/cms/slug-of-record-556`, without any paths, since it's not going to improve dramatically your SEO, while yes your maintenance burden.

Instead of that, you need to put breadcrumbs in the website. For that, check:

- [How to make them with bootstrap](http://getbootstrap.com/components/#breadcrumbs).
- [How to use schema.org markup to make Google interpret those breadcrumbs automatically](https://developers.google.com/search/docs/data-types/breadcrumbs).
- Example 2 in above link.